File: src/minitorch/tensor.py
"""Tensor：张量前端对象（Ch1）。

持有 shape/stride/storage_offset + Storage 引用。
view 系列操作只改元信息不拷贝；reshape 在非连续时 materialize。
算子为纯计算（无 autograd），自动微分在 Ch2/Ch3 引入。
对应真实 PyTorch 的 at::Tensor / c10::TensorImpl。
"""

from __future__ import annotations

from collections.abc import Sequence

import numpy as np

from .storage import Storage


def _prod(shape: Sequence[int]) -> int:
    n = 1
    for d in shape:
        n *= d
    return n


def _compute_contiguous_strides(shape: Sequence[int]) -> tuple[int, ...]:
    strides = [1] * len(shape)
    for i in range(len(shape) - 2, -1, -1):
        strides[i] = strides[i + 1] * shape[i + 1]
    return tuple(strides)


def _infer_shape(shape: Sequence[int], numel: int) -> tuple[int, ...]:
    shape = list(shape)
    if -1 in shape:
        if shape.count(-1) != 1:
            raise ValueError("view/reshape: 只能有一个 -1")
        known = _prod([d for d in shape if d != -1])
        shape[shape.index(-1)] = numel // known
    return tuple(shape)


def _broadcast_shapes(*shapes: tuple[int, ...]) -> tuple[int, ...]:
    if not shapes:
        return ()
    ndim = max(len(s) for s in shapes)
    aligned = [(1,) * (ndim - len(s)) + s for s in shapes]
    result = []
    for dims in zip(*aligned, strict=True):
        non_one = [d for d in dims if d != 1]
        if not non_one:
            result.append(1)
        elif len(set(non_one)) == 1:
            result.append(non_one[0])
        else:
            raise ValueError(f"shape 无法广播: {shapes}")
    return tuple(result)


class Tensor:
    """张量：Storage 的视图，持有 shape/strides/storage_offset。"""

    def __init__(
        self,
        storage: Storage,
        shape: Sequence[int],
        strides: Sequence[int],
        storage_offset: int = 0,
        requires_grad: bool = False,
    ):
        self._storage = storage
        self._shape = tuple(shape)
        self._strides = tuple(strides)
        self._storage_offset = storage_offset
        self.requires_grad = requires_grad
        self.grad: Tensor | None = None
        self.grad_fn = None

    @property
    def shape(self) -> tuple[int, ...]:
        return self._shape

    @property
    def strides(self) -> tuple[int, ...]:
        return self._strides

    @property
    def storage_offset(self) -> int:
        return self._storage_offset

    @property
    def storage(self) -> Storage:
        return self._storage

    @property
    def dtype(self) -> np.dtype:
        return self._storage.dtype

    @property
    def ndim(self) -> int:
        return len(self._shape)

    @property
    def size(self) -> int:
        return _prod(self._shape)

    @property
    def T(self) -> Tensor:
        if self.ndim < 2:
            return self
        return self.transpose()

    def numel(self) -> int:
        return self.size

    def dim(self) -> int:
        return self.ndim

    @classmethod
    def from_numpy(cls, arr: np.ndarray) -> Tensor:
        arr = np.asarray(arr)
        storage = Storage.from_numpy(arr)
        return cls(storage, arr.shape, _compute_contiguous_strides(arr.shape))

    @classmethod
    def zeros(cls, *shape: int, dtype=np.float64) -> Tensor:
        return cls.from_numpy(np.zeros(shape, dtype=dtype))

    @classmethod
    def ones(cls, *shape: int, dtype=np.float64) -> Tensor:
        return cls.from_numpy(np.ones(shape, dtype=dtype))

    @classmethod
    def arange(cls, stop: float, start: float = 0, step: float = 1, dtype=np.float64) -> Tensor:
        return cls.from_numpy(np.arange(start, stop, step, dtype=dtype))

    @classmethod
    def randn(cls, *shape: int, dtype=np.float64) -> Tensor:
        return cls.from_numpy(np.random.randn(*shape).astype(dtype))

    @classmethod
    def rand(cls, *shape: int, dtype=np.float64) -> Tensor:
        return cls.from_numpy(np.random.rand(*shape).astype(dtype))

    def _numpy_view(self) -> np.ndarray:
        base = self._storage.data
        byte_strides = tuple(s * base.itemsize for s in self._strides)
        return np.lib.stride_tricks.as_strided(
            base[self._storage_offset:],
            shape=self._shape if self._shape else (),
            strides=byte_strides if self._shape else (),
        )

    def numpy(self) -> np.ndarray:
        return self._numpy_view().copy()

    def is_contiguous(self) -> bool:
        if self.ndim <= 1:
            return True
        return self._strides == _compute_contiguous_strides(self._shape)

    def contiguous(self) -> Tensor:
        if self.is_contiguous():
            return self
        arr = self._numpy_view().copy()
        storage = Storage.from_numpy(arr)
        return Tensor(storage, arr.shape, _compute_contiguous_strides(arr.shape), 0, self.requires_grad)

    def view(self, *shape: int) -> Tensor:
        shape = _infer_shape(shape, self.size)
        if not self.is_contiguous():
            raise RuntimeError("view 要求 contiguous，请用 .reshape() 或先 .contiguous()")
        if _prod(shape) != self.size:
            raise RuntimeError(f"view: 元素数不匹配 {shape} vs {self.size}")
        return Tensor(
            self._storage, shape, _compute_contiguous_strides(shape), self._storage_offset, self.requires_grad
        )

    def reshape(self, *shape: int) -> Tensor:
        shape = _infer_shape(shape, self.size)
        if _prod(shape) != self.size:
            raise RuntimeError("reshape: 元素数不匹配")
        if self.is_contiguous():
            return self.view(*shape)
        return self.contiguous().view(*shape)

    def _transpose(self, dim0: int = 1, dim1: int = 0) -> Tensor:
        if dim0 < 0:
            dim0 += self.ndim
        if dim1 < 0:
            dim1 += self.ndim
        shape = list(self._shape)
        strides = list(self._strides)
        shape[dim0], shape[dim1] = shape[dim1], shape[dim0]
        strides[dim0], strides[dim1] = strides[dim1], strides[dim0]
        return Tensor(self._storage, shape, strides, self._storage_offset, self.requires_grad)

    def transpose(self, dim0: int = 1, dim1: int = 0) -> Tensor:
        from .ops.arithmetic import Transpose

        return Transpose.apply(self, dim0=dim0, dim1=dim1)

    def permute(self, *dims: int) -> Tensor:
        dims = tuple(d + self.ndim if d < 0 else d for d in dims)
        if sorted(dims) != list(range(self.ndim)):
            raise RuntimeError("permute: dims 必须是 0..ndim-1 的排列")
        shape = tuple(self._shape[d] for d in dims)
        strides = tuple(self._strides[d] for d in dims)
        return Tensor(self._storage, shape, strides, self._storage_offset, self.requires_grad)

    def unsqueeze(self, dim: int) -> Tensor:
        if dim < 0:
            dim += self.ndim + 1
        shape = self._shape[:dim] + (1,) + self._shape[dim:]
        strides = self._strides[:dim] + (0,) + self._strides[dim:]
        return Tensor(self._storage, shape, strides, self._storage_offset, self.requires_grad)

    def squeeze(self, dim: int | None = None) -> Tensor:
        if dim is not None:
            if dim < 0:
                dim += self.ndim
            if self._shape[dim] != 1:
                return self
            shape = self._shape[:dim] + self._shape[dim + 1 :]
            strides = self._strides[:dim] + self._strides[dim + 1 :]
            return Tensor(self._storage, shape, strides, self._storage_offset, self.requires_grad)
        pairs = [(s, st) for s, st in zip(self._shape, self._strides, strict=True) if s != 1]
        if not pairs:
            return Tensor(self._storage, (), (), self._storage_offset, self.requires_grad)
        shape, strides = (tuple(p[0] for p in pairs), tuple(p[1] for p in pairs))
        return Tensor(self._storage, shape, strides, self._storage_offset, self.requires_grad)

    def broadcast_to(self, shape: Sequence[int]) -> Tensor:
        shape = tuple(shape)
        if len(shape) < self.ndim:
            raise RuntimeError("broadcast_to: 目标维度不能少于自身")
        pad = len(shape) - self.ndim
        self_shape = (1,) * pad + self._shape
        self_strides = (0,) * pad + self._strides
        new_strides = []
        for i, (s, t) in enumerate(zip(self_shape, shape, strict=True)):
            if s == t:
                new_strides.append(self_strides[i])
            elif s == 1:
                new_strides.append(0)
            else:
                raise RuntimeError(f"broadcast_to: {self._shape} 不能广播到 {shape}")
        return Tensor(self._storage, shape, tuple(new_strides), self._storage_offset, self.requires_grad)

    @staticmethod
    def broadcast_tensors(*tensors: Tensor) -> tuple[Tensor, ...]:
        out_shape = _broadcast_shapes(*[t.shape for t in tensors])
        return tuple(t.broadcast_to(out_shape) for t in tensors)

    def __getitem__(self, key) -> Tensor:
        arr = self._numpy_view()[key]
        return Tensor.from_numpy(np.asarray(arr))

    def __setitem__(self, key, value):
        arr = self._numpy_view()
        if isinstance(value, Tensor):
            value = value._numpy_view()
        arr[key] = value

    def _ensure_tensor(self, other) -> Tensor:
        if isinstance(other, Tensor):
            return other
        return Tensor.from_numpy(np.asarray(other))

    def _binary(self, other, op) -> Tensor:
        other = self._ensure_tensor(other)
        a, b = Tensor.broadcast_tensors(self, other)
        return Tensor.from_numpy(op(a._numpy_view(), b._numpy_view()))

    def _add(self, other) -> Tensor:
        return self._binary(other, lambda a, b: a + b)

    def _sub(self, other) -> Tensor:
        return self._binary(other, lambda a, b: a - b)

    def _mul(self, other) -> Tensor:
        return self._binary(other, lambda a, b: a * b)

    def _div(self, other) -> Tensor:
        return self._binary(other, lambda a, b: a / b)

    def _pow(self, other) -> Tensor:
        return self._binary(other, lambda a, b: a**b)

    def _neg(self) -> Tensor:
        return Tensor.from_numpy(-self._numpy_view())

    def _matmul(self, other) -> Tensor:
        other = self._ensure_tensor(other)
        return Tensor.from_numpy(self._numpy_view() @ other._numpy_view())

    def _sum(self, dim: int | None = None, keepdim: bool = False) -> Tensor:
        arr = self._numpy_view().sum(axis=dim, keepdims=keepdim)
        return Tensor.from_numpy(np.asarray(arr))

    def _mean(self, dim: int | None = None, keepdim: bool = False) -> Tensor:
        arr = self._numpy_view().mean(axis=dim, keepdims=keepdim)
        return Tensor.from_numpy(np.asarray(arr))

    def max(self, dim: int | None = None):
        arr = self._numpy_view()
        if dim is None:
            return arr.max()
        return Tensor.from_numpy(arr.max(axis=dim))

    def add(self, other) -> Tensor:
        from .ops.arithmetic import Add

        return Add.apply(self, self._ensure_tensor(other))

    def sub(self, other) -> Tensor:
        from .ops.arithmetic import Sub

        return Sub.apply(self, self._ensure_tensor(other))

    def mul(self, other) -> Tensor:
        from .ops.arithmetic import Mul

        return Mul.apply(self, self._ensure_tensor(other))

    def div(self, other) -> Tensor:
        from .ops.arithmetic import Div

        return Div.apply(self, self._ensure_tensor(other))

    def pow(self, other) -> Tensor:
        from .ops.arithmetic import Pow

        return Pow.apply(self, other)

    def neg(self) -> Tensor:
        from .ops.arithmetic import Neg

        return Neg.apply(self)

    def matmul(self, other) -> Tensor:
        from .ops.arithmetic import Matmul

        return Matmul.apply(self, self._ensure_tensor(other))

    def sum(self, dim: int | None = None, keepdim: bool = False) -> Tensor:
        from .ops.arithmetic import Sum

        return Sum.apply(self, dim=dim, keepdim=keepdim)

    def mean(self, dim: int | None = None, keepdim: bool = False) -> Tensor:
        from .ops.arithmetic import Mean

        return Mean.apply(self, dim=dim, keepdim=keepdim)

    def __add__(self, other): return self.add(other)
    def __radd__(self, other): return self._ensure_tensor(other).add(self)
    def __sub__(self, other): return self.sub(other)
    def __rsub__(self, other): return self._ensure_tensor(other).sub(self)
    def __mul__(self, other): return self.mul(other)
    def __rmul__(self, other): return self._ensure_tensor(other).mul(self)
    def __truediv__(self, other): return self.div(other)
    def __rtruediv__(self, other): return self._ensure_tensor(other).div(self)
    def __pow__(self, other): return self.pow(other)
    def __matmul__(self, other): return self.matmul(other)
    def __neg__(self): return self.neg()

    def item(self):
        return self._numpy_view().item()

    def tolist(self):
        return self._numpy_view().tolist()

    def backward(
        self,
        gradient: Tensor | None = None,
        retain_graph: bool = False,
        retain_grad: bool = False,
    ) -> None:
        from .autograd.variable import backward

        backward(self, gradient, retain_graph=retain_graph, retain_grad=retain_grad)

    def allclose(self, other, atol: float = 1e-8) -> bool:
        other = self._ensure_tensor(other)
        return bool(np.allclose(self._numpy_view(), other._numpy_view(), atol=atol))

    def __len__(self) -> int:
        return self._shape[0] if self._shape else 0

    def __repr__(self) -> str:
        return f"Tensor({self._numpy_view().tolist()}, shape={self._shape})"

File: cpp/c10/tensor.cpp
// TensorImpl 实现（Ch8）

#include "c10/tensor.h"
#include "autograd/engine.h"
#include "autograd/grad_mode.h"
#include <sstream>
#include <stdexcept>
#include <algorithm>
#include <cmath>

namespace minitorch {

// ── 工具函数 ─────────────────────────────────────────

std::vector<int64_t> compute_contiguous_strides(const std::vector<int64_t>& shape) {
    int64_t n = static_cast<int64_t>(shape.size());
    std::vector<int64_t> strides(n);
    if (n == 0) return strides;
    strides[n - 1] = 1;
    for (int64_t i = n - 2; i >= 0; --i) {
        strides[i] = strides[i + 1] * shape[i + 1];
    }
    return strides;
}

std::vector<int64_t> infer_shape(std::vector<int64_t> shape, int64_t numel) {
    int64_t known = 1;
    int neg_count = 0;
    for (auto& d : shape) {
        if (d == -1) ++neg_count;
        else known *= d;
    }
    if (neg_count > 1) throw std::runtime_error("只能有一个 -1");
    if (neg_count == 1) {
        for (auto& d : shape) {
            if (d == -1) d = numel / known;
        }
    }
    return shape;
}

std::vector<int64_t> broadcast_shapes(const std::vector<int64_t>& a,
                                       const std::vector<int64_t>& b) {
    int64_t na = static_cast<int64_t>(a.size());
    int64_t nb = static_cast<int64_t>(b.size());
    int64_t n = std::max(na, nb);
    std::vector<int64_t> result(n);
    for (int64_t i = 0; i < n; ++i) {
        int64_t da = (i < na) ? a[na - 1 - i] : 1;
        int64_t db = (i < nb) ? b[nb - 1 - i] : 1;
        if (da == 1) result[n - 1 - i] = db;
        else if (db == 1) result[n - 1 - i] = da;
        else if (da == db) result[n - 1 - i] = da;
        else throw std::runtime_error("shape 无法广播");
    }
    return result;
}

int64_t numel(const std::vector<int64_t>& shape) {
    int64_t n = 1;
    for (auto d : shape) n *= d;
    return n;
}

// ── TensorImpl 构造 ──────────────────────────────────

TensorImpl::TensorImpl(std::shared_ptr<Storage> storage,
                       std::vector<int64_t> shape,
                       std::vector<int64_t> strides,
                       int64_t storage_offset,
                       bool requires_grad)
    : storage_(storage), shape_(std::move(shape)), strides_(std::move(strides)),
      storage_offset_(storage_offset), requires_grad_(requires_grad) {}

TensorImpl::TensorImpl(const std::vector<double>& data,
                       std::vector<int64_t> shape,
                       bool requires_grad)
    : shape_(std::move(shape)),
      strides_(compute_contiguous_strides(shape_)),
      storage_offset_(0),
      requires_grad_(requires_grad) {
    storage_ = make_storage(data);
}

TensorImpl::TensorImpl(const TensorImpl& other)
    : storage_(other.storage_), shape_(other.shape_), strides_(other.strides_),
      storage_offset_(other.storage_offset_), requires_grad_(other.requires_grad_) {}

TensorImpl& TensorImpl::operator=(const TensorImpl& other) {
    if (this != &other) {
        storage_ = other.storage_;
        shape_ = other.shape_;
        strides_ = other.strides_;
        storage_offset_ = other.storage_offset_;
        requires_grad_ = other.requires_grad_;
    }
    return *this;
}

// ── 形状操作 ─────────────────────────────────────────

bool TensorImpl::is_contiguous() const {
    if (numel() <= 1) return true;
    return strides_ == compute_contiguous_strides(shape_);
}

TensorImplPtr TensorImpl::contiguous() const {
    if (is_contiguous()) {
        return std::make_shared<TensorImpl>(*this);
    }
    // materialize：按 stride 读取数据到连续 buffer
    auto n = numel();
    std::vector<double> flat(static_cast<size_t>(n));
    // 遍历所有逻辑索引
    std::vector<int64_t> indices(ndim(), 0);
    for (int64_t i = 0; i < n; ++i) {
        flat[static_cast<size_t>(i)] = storage_->data()[static_cast<size_t>(linear_offset(indices))];
        // 递增索引（行优先）
        for (int64_t d = ndim() - 1; d >= 0; --d) {
            if (++indices[d] < shape_[d]) break;
            indices[d] = 0;
        }
    }
    auto new_storage = make_storage(flat);
    return std::make_shared<TensorImpl>(new_storage, shape_,
                                        compute_contiguous_strides(shape_), 0, requires_grad_);
}

TensorImplPtr TensorImpl::view(std::vector<int64_t> shape) const {
    shape = infer_shape(std::move(shape), numel());
    if (!is_contiguous()) {
        throw std::runtime_error("view 要求 contiguous，请用 reshape() 或先 contiguous()");
    }
    auto new_strides = compute_contiguous_strides(shape);
    return std::make_shared<TensorImpl>(storage_, shape, new_strides,
                                        storage_offset_, requires_grad_);
}

TensorImplPtr TensorImpl::reshape(std::vector<int64_t> shape) const {
    shape = infer_shape(std::move(shape), numel());
    if (is_contiguous()) {
        return view(shape);
    }
    return contiguous()->view(shape);
}

TensorImplPtr TensorImpl::transpose(int64_t dim0, int64_t dim1) const {
    if (dim0 < 0) dim0 += ndim();
    if (dim1 < 0) dim1 += ndim();
    auto shape = shape_;
    auto strides = strides_;
    std::swap(shape[dim0], shape[dim1]);
    std::swap(strides[dim0], strides[dim1]);
    return std::make_shared<TensorImpl>(storage_, shape, strides,
                                        storage_offset_, requires_grad_);
}

TensorImplPtr TensorImpl::permute(std::vector<int64_t> dims) const {
    int64_t n = ndim();
    for (auto& d : dims) {
        if (d < 0) d += n;
    }
    std::vector<int64_t> new_shape, new_strides;
    for (auto d : dims) {
        new_shape.push_back(shape_[d]);
        new_strides.push_back(strides_[d]);
    }
    return std::make_shared<TensorImpl>(storage_, new_shape, new_strides,
                                        storage_offset_, requires_grad_);
}

// ── 数据访问 ─────────────────────────────────────────

int64_t TensorImpl::linear_offset(const std::vector<int64_t>& indices) const {
    int64_t offset = storage_offset_;
    for (size_t i = 0; i < indices.size() && i < strides_.size(); ++i) {
        offset += indices[i] * strides_[i];
    }
    return offset;
}

double TensorImpl::item() const {
    if (numel() != 1) {
        throw std::runtime_error("item() 要求 numel()==1");
    }
    return storage_->data()[static_cast<size_t>(storage_offset_)];
}

std::vector<double> TensorImpl::to_vector() const {
    auto n = numel();
    std::vector<double> result(static_cast<size_t>(n));
    std::vector<int64_t> indices(ndim(), 0);
    for (int64_t i = 0; i < n; ++i) {
        result[static_cast<size_t>(i)] = storage_->data()[static_cast<size_t>(linear_offset(indices))];
        for (int64_t d = ndim() - 1; d >= 0; --d) {
            if (++indices[d] < shape_[d]) break;
            indices[d] = 0;
        }
    }
    return result;
}

void TensorImpl::from_vector(const std::vector<double>& data) {
    if (static_cast<int64_t>(data.size()) != numel()) {
        throw std::runtime_error("from_vector: 大小不匹配");
    }
    if (!is_contiguous()) {
        // 非连续：按索引写入
        std::vector<int64_t> indices(ndim(), 0);
        for (int64_t i = 0; i < numel(); ++i) {
            storage_->data()[static_cast<size_t>(linear_offset(indices))] = data[static_cast<size_t>(i)];
            for (int64_t d = ndim() - 1; d >= 0; --d) {
                if (++indices[d] < shape_[d]) break;
                indices[d] = 0;
            }
        }
    } else {
        // 连续：直接拷贝
        std::copy(data.begin(), data.end(),
                  storage_->data() + storage_offset_);
    }
}

void TensorImpl::set_item(const std::vector<int64_t>& indices, double value) {
    storage_->data()[static_cast<size_t>(linear_offset(indices))] = value;
}

TensorImplPtr TensorImpl::clone() const {
    auto data = to_vector();
    auto result = make_tensor(data, shape_);
    result->set_requires_grad(requires_grad_);
    return result;
}

TensorImplPtr TensorImpl::expand(std::vector<int64_t> target_shape) const {
    int64_t src_ndim = static_cast<int64_t>(shape_.size());
    int64_t dst_ndim = static_cast<int64_t>(target_shape.size());
    if (dst_ndim < src_ndim) {
        throw std::runtime_error("expand: target ndim < source ndim");
    }
    int64_t pad = dst_ndim - src_ndim;
    std::vector<int64_t> padded_shape(dst_ndim, 1);
    std::vector<int64_t> padded_strides(dst_ndim, 0);
    for (int64_t i = 0; i < src_ndim; ++i) {
        padded_shape[pad + i] = shape_[i];
        padded_strides[pad + i] = strides_[i];
    }
    std::vector<int64_t> new_strides(dst_ndim);
    for (int64_t i = 0; i < dst_ndim; ++i) {
        if (padded_shape[i] == target_shape[i]) {
            new_strides[i] = padded_strides[i];
        } else if (padded_shape[i] == 1) {
            new_strides[i] = 0;
        } else {
            throw std::runtime_error("expand: shape 不兼容");
        }
    }
    return std::make_shared<TensorImpl>(storage_, target_shape, new_strides,
                                        storage_offset_, requires_grad_);
}

void TensorImpl::fill_(double value) {
    int64_t n = numel();
    std::vector<int64_t> indices(ndim(), 0);
    for (int64_t i = 0; i < n; ++i) {
        storage_->data()[static_cast<size_t>(linear_offset(indices))] = value;
        for (int64_t d = ndim() - 1; d >= 0; --d) {
            if (++indices[d] < shape_[d]) break;
            indices[d] = 0;
        }
    }
}

void TensorImpl::zero_() {
    fill_(0.0);
}

std::string TensorImpl::repr() const {
    std::ostringstream oss;
    oss << "Tensor(shape=[";
    for (size_t i = 0; i < shape_.size(); ++i) {
        if (i > 0) oss << ", ";
        oss << shape_[i];
    }
    oss << "], data=[";
    auto data = to_vector();
    for (size_t i = 0; i < data.size() && i < 10; ++i) {
        if (i > 0) oss << ", ";
        oss << data[i];
    }
    if (data.size() > 10) oss << ", ...";
    oss << "])";
    return oss.str();
}

// ── backward 入口 ───────────────────────────────────

void TensorImpl::backward(TensorImplPtr gradient,
                          bool retain_graph,
                          bool retain_grad,
                          bool create_graph) {
    if (!grad_fn_) {
        throw std::runtime_error(
            "backward() called on a tensor with no grad_fn");
    }
    if (!gradient) {
        if (numel() != 1) {
            throw std::runtime_error(
                "grad can be implicitly created only for scalar outputs");
        }
        std::vector<double> ones(static_cast<size_t>(numel()), 1.0);
        gradient = make_tensor(ones, shape_);
    }
    run_backward(grad_fn_, gradient, retain_graph, retain_grad, create_graph);
    if (!retain_graph && !create_graph) {
        grad_fn_.reset();
    }
}

} // namespace minitorch